src/ILCompiler.Compiler/src/Compiler/ExplicitClassConstructors.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+using Internal.TypeSystem;
+
+using AssemblyName = System.Reflection.AssemblyName;
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    /// <summary>
+    /// Encapsulates a list of class constructors that must be run in a prescribed order during start-up
+    /// </summary>
+    public sealed class ExplicitClassConstructors
+    {
+        private static readonly IList<OrderedClassConstructorInfo> OrderedClassConstructors =
+            new ReadOnlyCollection<OrderedClassConstructorInfo>(new[]
+            {
+                new OrderedClassConstructorInfo("System.Private.CoreLib", "System.PreallocatedOutOfMemoryException"),
+                new OrderedClassConstructorInfo("System.Private.CoreLib", "System.Runtime.CompilerServices.ClassConstructorRunner+Cctor"),
+                new OrderedClassConstructorInfo("System.Private.CoreLib", "System.Runtime.CompilerServices.ClassConstructorRunner"),
```

Comment:
With the explicit Initialize method, we should be able to simplify this even furher: We can shrink all the CoreLib types into one that will call the random CoreLib initializations in the right order - so that we won't need to have the random list of CoreLib types hardcoded in the compiler.